packages/next/src/lib/metadata/resolve-metadata.ts

```diff
@@ -236,52 +237,81 @@ async function resolveStaticMetadata(components: ComponentsType) {
-    const [metadataExport, staticFilesMetadata] = item
-    const currentMetadata =
-      typeof metadataExport === 'function'
-        ? metadataExport(parentPromise)
-        : metadataExport
-    const layerMetadataPromise =
-      currentMetadata instanceof Promise
-        ? currentMetadata
-        : Promise.resolve(currentMetadata)
+  // call each `generateMetadata function concurrently and stash their resolver
+  for (let i = 0; i < metadataItems.length; i++) {
+    const { metadataExport } = metadataItems[i]
+    if (typeof metadataExport === 'function') {
+      generateMetadataResults.push(
+        metadataExport(
+          new Promise((resolve) => {
+            resolvers.push(resolve)
+          })
+        )
+      )
+    }
+  }
 
-    parentPromise = parentPromise.then((resolved) => {
-      return layerMetadataPromise.then((metadata) => {
-        resolved = deepClone(resolved)
-        merge(resolved, metadata, staticFilesMetadata, {
-          title: resolved.title?.template || null,
-          openGraph: resolved.openGraph?.title?.template || null,
-          twitter: resolved.twitter?.title?.template || null,
-        })
-        return process.env.NODE_ENV === 'development'
-          ? Object.freeze(resolved)
-          : resolved
-      })
+  // Loop over all metadata items again, merging synchronously any static object exports,
+  // awaiting any static promise exports, and resolving parent metadata and awaiting any generated
+  // metadata
+  let resolvingIndex = 0
+  for (let i = 0; i < metadataItems.length; i++) {
+    const { metadataExport, staticFilesMetadata, isLeafLayout } =
+      metadataItems[i]
+    let metadata: Metadata | null = null
+    if (typeof metadataExport === 'function') {
+      const resolveParent = resolvers[resolvingIndex]
+      const generatedMetadata = generateMetadataResults[resolvingIndex]
+      resolvingIndex++
+      // In dev we clone and freeze to prevent relying on mutating resolvedMetadata directly.
+      // In prod we just pass resolvedMetadata through without any copying.
+      const currentResolvedMetadata: ResolvedMetadata =
+        process.env.NODE_ENV === 'development'
+          ? Object.freeze(
+              require('next/dist/compiled/@edge-runtime/primitives/structured-clone').structuredClone(
+                resolvedMetadata
+              )
+            )
+          : resolvedMetadata
+
+      // This resolve should unblock the generateMetadata function if it awaited the parent
+      // argument. If it didn't await the parent argument it might already have a value since it was
+      // called concurrently. Regardless we await the return value before continuing on to the next
+      // layer
+      resolveParent(currentResolvedMetadata)
+      metadata = await generatedMetadata
+    } else {
+      metadata = metadataExport
+    }
+
+    // If the layout is the same layer with page, skip the
+    if (isLeafLayout && metadata) {
+      delete metadata.title
+    }
```

Comment:
I don't think this quite works. If there is no metadata defined in the Page or if the metadata does not specify a title then the leaf Layout title should be used. The way I would probably solve this is track the templates in a variable outside the for loop. update it on each iteration, reading from the resolvedMetadata. Then we skip this read if the last merged metadata was the leaf metadata